// honi-cf/dist/workflow.js
export function step(config, handler) {
    return { config, handler };
}
export function workflow(config) {
    // Return a class that can be used as a Cloudflare Workflow entrypoint
    // Use abstract class pattern to avoid needing WorkflowEntrypoint at runtime
    class HoniWorkflow {
        ctx;
        env;
        constructor(ctx, env) {
            this.ctx = ctx;
            this.env = env;
        }
        async run(event, step) {
            let lastResult = event.payload;
            try {
                for (const stepDef of config.steps) {
                    lastResult = await stepDef.handler(lastResult, step);
                }
                if (config.onComplete) {
                    await config.onComplete(lastResult, this.env);
                }
            }
            catch (err) {
                if (config.onError) {
                    await config.onError(err, this.env);
                }
                else {
                    throw err;
                }
            }
        }
    }
    return HoniWorkflow;
}
